### PR TITLE
Add mlp sklearn024

### PR DESCRIPTION
**Pull request by KEggensperger**

### CODE REVIEW

File: autosklearn/pipeline/components/regression/mlp.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+            self.max_iter = int(self.max_iter)
+            self.hidden_layer_depth = int(self.hidden_layer_depth)
+            self.num_nodes_per_layer = int(self.num_nodes_per_layer)
+            self.hidden_layer_sizes = tuple(self.num_nodes_per_layer
+                                            for i in range(self.hidden_layer_depth))
+            self.activation = str(self.activation)
+            self.alpha = float(self.alpha)
+            self.learning_rate_init = float(self.learning_rate_init)
+            self.early_stopping = str(self.early_stopping)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/regression/mlp.py:77]: This 1st cast isn't there for classification -> I'm not sure which one is correct.

File: autosklearn/pipeline/components/regression/mlp.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            except ValueError:
+                self.batch_size = str(self.batch_size)
+
+            self.shuffle = check_for_bool(self.shuffle)
+            self.beta_1 = float(self.beta_1)
+            self.beta_2 = float(self.beta_2)
+            self.epsilon = float(self.epsilon)
+            self.beta_1 = float(self.beta_1)
+
+            self.max_iter = int(self.max_iter)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/regression/mlp.py:110]: This argument isn't casted for classification -> I'm not sure which one is correct.

File: autosklearn/pipeline/components/regression/mlp.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            self.estimator = MLPRegressor(
+                hidden_layer_sizes=self.hidden_layer_sizes,
+                activation=self.activation,
+                solver=self.solver,
+                alpha=self.alpha,
+                batch_size=self.batch_size,
+                learning_rate_init=self.learning_rate_init,
+                max_iter=n_iter,
+                shuffle=self.shuffle,
+                random_state=self.random_state,
```
Comment (mfeurer) [inline autosklearn/pipeline/components/regression/mlp.py:125]: There's a copy.copy() for classification around self.random_state. As it isn't here, maybe it's not necessary for classification?

File: autosklearn/pipeline/components/regression/mlp.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    @staticmethod
+    def get_hyperparameter_search_space(dataset_properties=None):
+        cs = ConfigurationSpace()
+        hidden_layer_depth = UniformIntegerHyperparameter(name="hidden_layer_depth",
+                                                          lower=1, upper=3, default_value=1)
+        num_nodes_per_layer = UniformIntegerHyperparameter(name="num_nodes_per_layer",
+                                                           lower=16, upper=264, default_value=32,
+                                                           log=True)
+        activation = CategoricalHyperparameter(name="activation", choices=['tanh', 'logistic'],
```
Comment (mfeurer) [inline autosklearn/pipeline/components/regression/mlp.py:190]: Do we really want to use the logistic activation function?

File: autosklearn/pipeline/components/classification/mlp.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                epsilon=self.epsilon,
+                # We do not use these, see comments below in search space
+                # momentum=self.momentum,
+                # nesterovs_momentum=self.nesterovs_momentum,
+                # power_t=self.power_t,
+                # learning_rate=self.learning_rate,
+                # max_fun=self.max_fun
+            )
+        else:
+            new_max_iter = min(self.max_iter-self.estimator.n_iter_, n_iter)
```
[the remark this answers is not in the document]
Comment (KEggensperger) [reply autosklearn/pipeline/components/classification/mlp.py:142]: yes

File: autosklearn/pipeline/components/regression/mlp.py (hunk 2 above)
Comment (KEggensperger) [reply autosklearn/pipeline/components/regression/mlp.py:110]: it is, but for regression is was casted twice -> done

File: autosklearn/pipeline/components/regression/mlp.py (hunk 4 above)
Comment (KEggensperger) [reply autosklearn/pipeline/components/regression/mlp.py:190]: That's what we decided some time ago. What would be better?

File: autosklearn/pipeline/components/regression/mlp.py (hunk 3 above)
Comment (KEggensperger) [reply autosklearn/pipeline/components/regression/mlp.py:125]: This is needed for the idempotent test, which only exists for classification. Fitting the MLP changes self.random_state, so subsequent calls to fit will use a different weight initialization. Copying the random_state guarantees that fit() will always do the same

**Review by mfeurer — APPROVED** [review]

**Pull request merged**